File: May-LeetCoding-Challange-2020/Week4/451-sort-characters-by-frequency.py
class Solution:
    # Sol 1. Using HashMap
    def frequencySort(self, s: str) -> str:
        # edge case
        if not s:
            return s
        c = Counter(list(s))
        c_list = sorted(c, key=lambda x:c[x], reverse=True)
        
        output = []
        for char in c_list:
            output.append(char*c[char])
        
        return "".join(output)
    
    # Time Complexity: O(n) + O(nlogn) = O(nlogn)
    # Space Complexity: O(n)
    # Run Code Runtime: 44 ms
    # Submit Runtime: 44 ms, 64.02% faster
    # Submit Memory: 15.7 mb, 7.14% less 

File: May-LeetCoding-Challange-2020/Week4/451-Sort-Characters-By-Frequency.cpp
// Time Complexity: O(max length*count of characters used)
// Space Complexity: O(n)
// Runtime: 232 ms, faster than 8.75%
// Memory Usage: 9.1 MB, less than 100.00%
class Solution {
public:
    string frequencySort(string s) {
        vector<char> storage; //storaging each character once in order
        map<char, string> temp; //map for each character followed by appending characters
        
        int max = 0; //initialize max
        string result = ""; //empty string for appending strings in order from max to 1
        
        for(int i = 0; i < s.length(); i++){
            if(temp[s[i]] == "")
                storage.push_back(s[i]); //fill storage each character once
            
            temp[s[i]]+=s[i]; //append character to the same character string
            
            if(temp[s[i]].size() > max) //update max if larger
                max = temp[s[i]].size();
        }
        
        while(max > 0){ //from max to 1, concatenate strings in order
            for(int i = 0; i < storage.size(); i++){ //for each character, find strings with [max] length
                if(temp[storage[i]].size() == max){ //if max,
                    result += temp[storage[i]]; //append string to result
                }
            }
            
            max --; //decrement max
        }

        return result;
    }
};